[pzhengze] iP

zheng-ze opened the pull request:
# Duke
> "He who fails to plan is planning to fail." - Winston Churchill

Duke is a personal assistant that helps you keep track of your tasks. It's

- Text-based
- ~~Easy~~ REALLY EASY to use
- Just works

All you need to do to use Duke is,

1. Download it [here](https://github.com/pzhengze/iP/releases/latest).
2. Place it wherever
3. Open it
4. Add your tasks
5. Profit 📈 

Most importantly, it is **Free**!

Features:

- [x] Managing tasks
- [X] Searching tasks by name
- [ ] Searching tasks by time
- [ ] GUI interface
- [ ] Reminders

------

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Duke {
    public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
}
```

Commit:
Add Gradle support

Commit:
Finished Level 1

Commit:
Finished Level 2

Commit:
Created Task class

Commit:
Updated ToDoList class to work with Task class

Commit:
Finished Level 3

Commit:
Finished Level 4

Commit:
Finished Level 5

Commit:
Finished Level 6

Commit:
Fixed Level 5 bugs

Commit:
Finished A-TextUiTesting

Commit:
Shifted error handling to Duke class.

Commit:
Created new DukeException classes

Commit:
Added new inputs

Commit:
Added new DukeException

Commit:
Finished Level-7

Commit:
Added date and time support

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadlines.java
#	src/main/java/Events.java
#	src/main/java/TaskList.java

Commit:
Added Ui class to handle interactions with the user

Commit:
Added Parser Class to parse user input

Commit:
Added Storage Class to handle read/write of save file

Commit:
Cleaned up unused imports

Commit:
Added duke package

Commit:
Fixed duke package

Commit:
Merged with add-gradle-support branch

Commit:
Bug fix

Commit:
Added some unit testing

Commit:
Modified mainClassName in build.gradle

Commit:
Added JavaDocs

Commit:
Modified code to fit coding standard

Commit:
Added find functionality

Commit:
Merge branch 'branch-A-Javadoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
Fix CheckStyle errors.

Commit:
Merge branch 'branch-A-CheckStyle'

euph00 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,70 @@
+package duke;
+
+import duke.exceptions.DukeException;
```
good use of minimal importing, in rest of project as well

euph00 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,68 @@
+        }
+
+        // Interprets the extracted function and executes the correct TaskList method.
+        switch(fn) {
```
correctly indented switch case according to coding standard

Alvozyn replied:
Good use of switch code here! :)

Commit:
Add Level-10

Saha9616 commented on src/main/java/duke/exceptions/DukeMissingDescriptionException.java:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Represents an exception that occurs when description is missing while creating a Task.
+ * @author pzhengze
+ */
```
good amount of classes used for exceptions

Saha9616 commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,279 @@
+                    || toIndex < fromIndex + 7
+                    || s.substring(fromIndex + 7, toIndex).isBlank()
+                    || s.substring(toIndex + 5).isBlank()) {
+                throw new DukeMissingEventDateException();
```
condition of if statement is too long, maybe you can reduce the size

Saha9616 reviewed (commented):
overall you have handled exceptions in a good way, code quality is also very good

Alvozyn commented on src/main/java/duke/exceptions/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+package duke.exceptions;
```
Nice use of separating all the exception into packages as it makes it easy for people to know what the error is. Well done!

Alvozyn commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,68 @@
+    public String parseAndExecute(String command) throws DukeUnknownCommandException, DukeException {
+        // Extracts the command the other details of the command
+        String[] inputArray = command.split(" ");
+        String fn = inputArray[0];
```
I know fn stands for function but perhaps naming of fn could be changed into more understandable and something that can better reflect what the string is referring to "eg cmd"

Alvozyn commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,279 @@
+        int num = TaskList.stringToInt(s);
+
+        String output = "\t Noted. I've removed this task:\n";
+        if (num < -1 || num >= tasks.size()) {
```
I was wondering why did you put "< -1" here and not "<= -1" since the command "delete -1" is considered as a invalid input?

Alvozyn reviewed (commented):
Overall, your code quality is really well done! The use of the switch cases are nicely implemented throughout the entire code. There were hardly any arrowheads in your code too making your code quality really clean and consistent! I definitely learn a lot from your code on how I can make improvements on my code quality. :)

Commit:
Bug Fix.

Commit:
Merge branch 'branch-Level-10'

Commit:
Fix checkstyle errors.

Commit:
Add assert

Added assert statements for assumptions made in code.

Commit:
Fix Code Quality

*Abstracted loadSaveFile in Storage.java and add in TaskList.java into smaller methods. They were too long.

*Fixed header comments for mark and unmark in Task.java

zheng-ze committed:
Merge pull request #2 from pzhengze/branch-A-Assertions

Add assert

zheng-ze committed:
Merge branch 'master' into branch-A-CodeQuality

zheng-ze committed:
Merge pull request #3 from pzhengze/branch-A-CodeQuality

Fix Code Quality

Commit:
Fix Bug

Cannot mark and unmark last item in list. Changed checks for valid range.

zheng-ze committed:
Create gradle.yml

Commit:
Add Better Search

*Find items even if keyword matches item only partially.
*Case-insensitive

Commit:
Add more abstraction for creation of Deadline and Event Objects.

Commit:
Add Rescalable GUI and chat boxes.

zheng-ze committed:
Add representative screenshot

zheng-ze committed:
Update README.md

zheng-ze committed:
Update README.md

zheng-ze committed:
Update README.md

Commit:
Fix Bug

Commit:
Fix checkstyle errors

zheng-ze committed:
Update README.md

zheng-ze committed:
Update README.md

zheng-ze committed:
Update README.md

zheng-ze committed:
Update README.md
